File: 第二次作业-贪吃蛇/GDIDemo2/Sprite.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;



/*******************************************************************************
	建立精灵类,用于管理精灵自身的行为,包括将图片渲染到指定的地方,
	并且能检查是否与其它精灵碰撞

*******************************************************************************/





namespace GameFramework2D
{
    class Sprite
    {

        protected string name;
        protected Bitmap texture;
        public Bitmap Texture
        {
            set
            {
                texture = value;
            }
            get
            {
                return texture;
            }
        }

        protected Point vPos;                   //位置向量

        public Point Pos															//设置精灵的位置
        {
            set
            {
                this.vPos = value;
                matTranslation.Reset();
                matTranslation.Translate(vPos.X - vHotSpot.X, vPos.Y - vHotSpot.Y);
                CalculateBoundingBox();
            }
            get
            {
                return vPos;
            }
        }
        
        protected int tx;																		//纹理图x
        protected int ty;																		//纹理图y									


        protected int width;																	//精灵宽度
        protected int height;																	//精灵高度			

        protected bool bVisible;
        protected bool bEnabled;

        protected Point vHotSpot;

        public Point HotSpot													//设置热点		
        {
            set
            {
                this.vHotSpot = value;
                matTranslation.Reset();
                matTranslation.Translate(vPos.X - vHotSpot.X, vPos.Y - vHotSpot.Y);
                CalculateBoundingBox();
            }
            get
            {
                return vHotSpot;
            }
        }
       

        protected Matrix matWorld;
        protected Matrix matScale;
        protected Matrix matRotation;
        protected Matrix matTranslation;



        protected float fRotationAngle;
        protected float fsx;
        protected float fsy;



        protected Point[] spriteBoundingBox;
        protected RectangleF calcBoundingBox;

        protected Pen boundingBoxPen;

        public Sprite(	Bitmap texture,														//纹理
                        Point vPos, 														//位置向量
                        int w = 0,																//精灵宽度
                        int h = 0,																//精灵高度
                        int tx = 0,															//纹理x
                        int ty = 0,															//纹理y
                        string name = "")
        {
           
            this.name = name;
            if (w == 0 || h == 0)
            {
                this.width = texture.Width;
                this.height = texture.Height;
            }
            else
            {
                this.width = w;
                  this.height = h;
            }
            this.tx = tx;
            this.ty = ty;
            this.texture = texture;
            spriteBoundingBox = new Point[4];
            this.bVisible = this.bEnabled = true;
            fRotationAngle = 0.0f;
            fsx = fsy = 0.0f;
            
           
            spriteBoundingBox[0].X = 0;
            spriteBoundingBox[0].Y = 0;
            spriteBoundingBox[1].X = 0;
            spriteBoundingBox[1].Y = h;
            spriteBoundingBox[2].X = w;
            spriteBoundingBox[2].Y = 0;
            spriteBoundingBox[3].X = w;
            spriteBoundingBox[3].Y = h;

            matTranslation = new Matrix();
            matRotation = new Matrix();
            matScale = new Matrix();
            matWorld = new Matrix();

            boundingBoxPen = new Pen(Color.Blue);
            Pos = vPos;
        }
       
        protected bool isDrawBoundingBox;
        public bool IsDrawBoudingBox
        {
            set
            {
                isDrawBoundingBox = value;
            }
            get
            {
                return isDrawBoundingBox;
            }
        }
   

        
        protected void ApplayTransform()
        {
            matWorld.Reset();

            matWorld.Multiply(matScale);
            matWorld.Multiply(matRotation, MatrixOrder.Append);
            matWorld.Multiply(matTranslation, MatrixOrder.Append);
            
        }
        protected void CalculateBoundingBox()
        {

            ApplayTransform();
        
            PointF[] ptTmp = new PointF[4];

            for (int i = 0; i < 4; ++i)
            {
                ptTmp[i] = spriteBoundingBox[i];
            }
            matWorld.TransformPoints(ptTmp);

            float fMinX = ptTmp[0].X;
            float fMaxX = ptTmp[0].X;
            float fMinY = ptTmp[0].Y;
            float fMaxY = ptTmp[0].Y;
            for (int i = 1; i < 4; ++i)
            {
                if (fMinX > ptTmp[i].X)
                {
                    fMinX = ptTmp[i].X;
                }
                if (fMaxX < ptTmp[i].X)
                {
                    fMaxX = ptTmp[i].X;
                }

                if (fMinY > ptTmp[i].Y)
                {
                    fMinY = ptTmp[i].Y;
                }
                if (fMaxY < ptTmp[i].Y)
                {
                    fMaxY = ptTmp[i].Y;
                }

            }
            calcBoundingBox.X = fMinX;
            calcBoundingBox.Y = fMinY;
            calcBoundingBox.Width = fMaxX - fMinX;
            calcBoundingBox.Height = fMaxY - fMinY;

        }

        public virtual void Update(float fDeltaTime) { }	//更新


        

        public bool IsSelected(int x, int y)
        {
            bool bRet = false;
            if (x >= calcBoundingBox.Left && x <= calcBoundingBox.Right && y >= calcBoundingBox.Top && y <= calcBoundingBox.Bottom)
            {
                return true;
            }
            return bRet;
        }
        public bool CollideWith(Sprite sprite)															//是否碰撞
        {
            return calcBoundingBox.IntersectsWith(sprite.calcBoundingBox);
        }
        public bool CollideWith(RectangleF rect)																//是否和某个矩形碰撞
        {
            return calcBoundingBox.IntersectsWith(rect);
        }

        public void SetScaleFactor(float sx, float sy)														//设置缩放系数 
        {
            fsx = sx;
            fsy = sy;
            matScale.Reset();
            matScale.Scale(fsx, fsy);

            CalculateBoundingBox();

        }

        public void SetDestRect( Point vPos, float w, float h)					//设置渲染的目的矩形
        {
            Pos = vPos;
            float sx = w / width;
            float sy = h / height;
            SetScaleFactor(sx, sy);
        }

        public void SetTexPos(int x, int y)																//设置精灵在纹理上截取的左上角坐标	
        {
            GraphicsUnit units = GraphicsUnit.Pixel;
            RectangleF rect = texture.GetBounds(ref units);

            if (x < 0 || x >= rect.Width || y < 0 || y >= rect.Height)
            {
                return;
            }
           
            tx = x;
            ty = y;
            
        }
        
        public void SetRotationAngle(float fAngle)														//设置旋转角度	
        {
            fRotationAngle = fAngle;
            matRotation.Reset();
            matRotation.RotateAt(fRotationAngle,vHotSpot);//, vHotSpot);
            CalculateBoundingBox();
        }
        public  float GetRotationAngle()
        {
            return fRotationAngle;
        }

      
        public bool IsVisible
        {
            set 
            {
                bVisible = value;
            }
            get
            {
                return bVisible;
            }
        }

        public bool IsEnable
        {
            set
            {
                bEnabled = value;
            }
            get
            {
                return bEnabled;
            }
        }

       

        public int GetWidth()																				//得到精灵宽度
        {
            return width;
        }
        public int GetHeight()																		//得到精灵高度
        {
            return height;
        }



        public string GetSpriteName()
        {
            return name;
        }
        public RectangleF GetBoundingBox()																//得到精灵矩形范围	
        {
            return calcBoundingBox;
        }

        public void Render(Graphics g)
        {
            g.Transform = matWorld;
            
            g.DrawImage(texture, tx, ty, width, height);
            if (isDrawBoundingBox)
            {
                g.ResetTransform();
                g.DrawRectangle(boundingBoxPen, Rectangle.Round(calcBoundingBox));
            }
            
        }
        public void Render(Graphics g,Point vPos)
        {
            Pos = vPos;
            Render(g);
        }
        
    }
}

File: 2015/DirectXWork2/DirectXWork2/框架-3CGame/Sprite.cpp

#include "Sprite.h"



CSprite::~CSprite(void)
{
	Release();
}
CSprite::CSprite(ID2D1HwndRenderTarget 	*pRenderTarget,													//RenderTarget													
	ID2D1Bitmap					*pTexture,																//纹理
	D2D1_POINT_2F				&vPos, 																	//位置向量
	int							w,																		//精灵宽度
	int							h,																		//精灵高度
	int							tx,																		//纹理x
	int							ty,																		//纹理y
	D2D1_POINT_2F				vHotSpot,																//热点
	wstring						wsName
	)
{
	assert(pRenderTarget != NULL);
	assert(pTexture != NULL);
	this->wsName = wsName;
	this->pTexture = pTexture;
	this->pRenderTarget = pRenderTarget;
	this->width = w;
	this->height = h;
	this->bStretch = true;

	D2D1_SIZE_F size=pTexture->GetSize();

	this->nTexWidth =  size.width;
	this->nTexHeight  = size.height;

	ptBoundingBox[0].x = 0;
	ptBoundingBox[0].y = 0;
	ptBoundingBox[1].x = 0;
	ptBoundingBox[1].y = h;
	ptBoundingBox[2].x = w;
	ptBoundingBox[2].y = 0;
	ptBoundingBox[3].x = w;
	ptBoundingBox[3].y = h;


	this->fRotationAngle = 0;
	this->vPos = vPos;
	this->tx = tx;
	this->ty = ty;
	this->vHotSpot = vHotSpot;
	this->bVisible = true;
	this->bEnabled = true;
	this->bTexPosChanged = false;

	pRenderTarget->CreateBitmap(D2D1::SizeU(width,height),D2D1::BitmapProperties(pTexture->GetPixelFormat()),&pSpriteBitmap);
	assert(pSpriteBitmap != NULL);
	pSpriteBitmap->CopyFromBitmap(&D2D1::Point2U(0,0),pTexture,&D2D1::RectU(tx,ty,tx + width,ty + height));
	pRenderTarget->CreateBitmapBrush(pSpriteBitmap,&pSpriteBrush);

	matWorld = Matrix3x2F::Identity();
	matScale = Matrix3x2F::Identity();
	matRotation = Matrix3x2F::Identity();
	matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x ,vPos.y - vHotSpot.y);
	matWorld = matScale * matRotation * matTranslation;
	this->ComputerBoundingBox();
	color = D2D1::ColorF(0xff,1.0f);
	pRenderTarget->CreateSolidColorBrush(color,&pBoundingBoxBrush);
	bDrawBoundingBox = false;//是否画包围盒

	rtBoundingBox.left = vPos.x;
	rtBoundingBox.top = vPos.y;
	rtBoundingBox.right = vPos.x + width;
	rtBoundingBox.bottom = vPos.y + height;

}

void CSprite::Render()
{
	if (!pRenderTarget)
	{
		return;
	}

	if (!bVisible)//可见性
		return;

	if (bStretch)
	{
		//此处代码填写完整
		matWorld = matScale * matRotation * matTranslation;
		pRenderTarget->SetTransform(matWorld);
		pRenderTarget->DrawBitmap(pSpriteBitmap);
		if (bDrawBoundingBox)
		{
			pRenderTarget->SetTransform(Matrix3x2F::Identity());
			pRenderTarget->DrawRectangle(rtBoundingBox,pBoundingBoxBrush,2.0f);
		}
	}
	else
	{
		Render(destRect,bStretch);
	}
}

void CSprite::Render(D2D1_POINT_2F &vPos)
{
	if (!pRenderTarget)
	{
		return;
	}

	if (!bVisible)//可见性
		return;

	this->vPos = vPos;
	matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x ,vPos.y - vHotSpot.y );
	matWorld =  matScale * matRotation * matTranslation;
	this->ComputerBoundingBox();
	Render();
}
void CSprite::Render(D2D1_RECT_F &destRect,bool bStretch,D2D1_EXTEND_MODE extendModeX, D2D1_EXTEND_MODE extendModeY)
{
	if (!pRenderTarget)
	{
		return;
	}
	
	if (!bVisible)//可见性
		return;

	if (bStretch)
	{
		fsx = (destRect.right - destRect.left) / static_cast<float>(width);
		fsy = (destRect.bottom - destRect.top ) / static_cast<float>(height);
		vPos.x = destRect.left;
		vPos.y = destRect.top;
		matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x ,vPos.y - vHotSpot.y );
		matScale = D2D1::Matrix3x2F::Scale(fsx,fsy,D2D1::Point2F(vHotSpot.x,vHotSpot.y));
		matWorld =  matScale * matRotation * matTranslation;
		this->ComputerBoundingBox();
		Render();
	}
	else
	{
		HRESULT hr = S_OK;
		

		if (this->bTexPosChanged)
		{
			SafeRelease(&pSpriteBrush);
			hr = pRenderTarget->CreateBitmapBrush(pSpriteBitmap,&pSpriteBrush);
			this->bTexPosChanged = false;
		}
		pSpriteBrush->SetExtendModeX(extendModeX);
		pSpriteBrush->SetExtendModeY(extendModeY);
		if (SUCCEEDED(hr))
		{
			matWorld = D2D1::Matrix3x2F::Identity();
			pRenderTarget->SetTransform(matWorld);//D2D1::Matrix3x2F::Identity());
			pRenderTarget->FillRectangle(&destRect,pSpriteBrush);
			if (bDrawBoundingBox)
			{
				pRenderTarget->SetTransform(Matrix3x2F::Identity());
				pRenderTarget->DrawRectangle(rtBoundingBox,pBoundingBoxBrush,2.0f);
			}		
		}
	}


}
bool CSprite::CollideWith(RECT &rect)
{
	RECT rtIntersect;
	RECT *rtSprite = reinterpret_cast<RECT*>(&rtBoundingBox);
	return IntersectRect(&rtIntersect,rtSprite,&rect);
}
bool CSprite::CollideWith(CSprite &sprite) 
{
	RECT rtIntersect;

	const RECT *rtSprite = reinterpret_cast<const RECT*>(sprite.GetBoundingBox());
	//return IntersectRect(&rtIntersect,rtSprite,&rtIntersect);//两个rtIntersect
	RECT *rtSpriteNew = reinterpret_cast<RECT*>(&rtBoundingBox);
	return IntersectRect(&rtIntersect, rtSprite, rtSpriteNew);
}

D2D1_POINT_2F &CSprite::GetHotSpot() 
{
	return this->vHotSpot;
}

const D2D1_RECT_F* CSprite::GetBoundingBox()
{
	return &rtBoundingBox;																	;
}																	

void CSprite::SetTexture(ID2D1Bitmap *tex)
{
	this->pTexture = tex;

}
const ID2D1Bitmap *CSprite::GetTexture()
{
	return pTexture;
}
const ID2D1Bitmap *CSprite::GetBitmap()
{
	return pSpriteBitmap;
}
void CSprite::SetHotSpot(D2D1_POINT_2F &vHotSpot)
{
	//请将此处代码填写完整
	this->vHotSpot.x = vHotSpot.x;
	this->vHotSpot.y = vHotSpot.y;
	matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x, vPos.y - vHotSpot.y);
	this->ComputerBoundingBox();

}

void CSprite::SetTransform(Matrix3x2F &matTrans)
{
	this->matWorld = matTrans;
	this->ComputerBoundingBox();

}
const Matrix3x2F *CSprite::GetTransform()
{
	return &matWorld;
}

int CSprite::GetWidth()
{
	return width;
}
int CSprite::GetHeight()
{
	return height;
}

void CSprite::SetPos(D2D1_POINT_2F &vPos)
{
	//请将此处代码填写完整
	this->vPos.x = vPos.x;
	this->vPos.y = vPos.y;
	matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x, vPos.y - vHotSpot.y);
	this->ComputerBoundingBox();
}
D2D1_POINT_2F &CSprite::GetPos()
{
	return vPos;
}
void CSprite::SetDestRect(float x1,float y1,float w,float h,bool bStretch) 
{

	fsx = (float)w / (float)width;
	fsy = (float)h / (float)height;
	vPos.x = x1  ;
	vPos.y = y1 ;
	matTranslation = Matrix3x2F::Translation(vPos.x - vHotSpot.x ,vPos.y - vHotSpot.y );
	SetScaleFactor(fsx,fsy);
	this->bStretch = bStretch;
	destRect.left = x1;
	destRect.top = y1;
	destRect.right = x1+w;
	destRect.bottom = y1 + h;
	if(!bStretch)
	{
		rtBoundingBox  = destRect;
	}

}
void CSprite::SetTexPos(int x,int y)
{
	assert(pSpriteBitmap != NULL);
	//请将此处代码填写完整
	D2D1_SIZE_F size = pTexture->GetSize();
	if (x >= size.width || x < 0 || y >= size.height || y < 0)
		return;
	pSpriteBitmap->CopyFromBitmap(&D2D1::Point2U(0, 0), pTexture, &D2D1::RectU(x, y, x + width, y + height));
	this->tx = x;
	this->ty = y;
}
void CSprite::Release()
{
	SafeRelease(&pSpriteBitmap);
	SafeRelease(&pSpriteBrush);
	SafeRelease(&pBoundingBoxBrush);
}



void CSprite::SetVisible(bool bVal)
{
	this->bVisible = bVal;
}
bool CSprite::IsVisible()
{
	return bVisible;
}

void CSprite::SetRotationAngle(float fAngle)															//设置旋转角度					
{
	//请将此处代码填写完整
	this->fRotationAngle = fAngle;
	matRotation = D2D1::Matrix3x2F::Rotation(fRotationAngle, D2D1::Point2F(vHotSpot.x,vHotSpot.y));
	ComputerBoundingBox();
}
float CSprite::GetRotationAngle()
{
	return fRotationAngle;
}
void CSprite::SetScaleFactor(float sx,float sy)
{
	//请将此处代码填写完整
	this->fsx = sx;
	this->fsy = sy;
	matScale = D2D1::Matrix3x2F::Scale(fsx, fsy, D2D1::Point2F(vHotSpot.x, vHotSpot.y));
	ComputerBoundingBox();
}

const wstring &CSprite::GetSpriteName()
{
	return wsName;
}

void CSprite::ComputerBoundingBox()
{
	
	//请将此处代码填写完整
	matWorld = matScale * matRotation * matTranslation;
	ptBoundingBox[0].x = 0;
	ptBoundingBox[0].y = 0;
	ptBoundingBox[1].x = 0;
	ptBoundingBox[1].y = height;
	ptBoundingBox[2].x = width;
	ptBoundingBox[2].y = 0;
	ptBoundingBox[3].x = width;
	ptBoundingBox[3].y = height;

	/*包围盒矩形*/
	for (int i = 0; i < 4; i++)
		ptBoundingBox[i] = matWorld.TransformPoint(ptBoundingBox[i]);


	/*精灵矩形*/
	int MaxX = ptBoundingBox[2].x, MaxY = ptBoundingBox[3].y;
	int MixX = ptBoundingBox[0].x, MixY = ptBoundingBox[1].y;
	for (int i = 0; i<4; i++)
	{
		if (ptBoundingBox[i].x >MaxX)
			MaxX = ptBoundingBox[i].x;
		if (ptBoundingBox[i].y >MaxY)
			MaxY = ptBoundingBox[i].y;
		if (ptBoundingBox[i].x <MixX)
			MixX = ptBoundingBox[i].x;
		if (ptBoundingBox[i].y <MixY)
			MixY = ptBoundingBox[i].y;
	}

	rtBoundingBox.left = MixX;
	rtBoundingBox.top = MixY;
	rtBoundingBox.right = MaxX;
	rtBoundingBox.bottom = MaxY;
}

bool CSprite::IsSelected(int x,int y)
{
	bool bRet = false;
	if (x>=rtBoundingBox.left && x <= rtBoundingBox.right  && y >= rtBoundingBox.top && y <= rtBoundingBox.bottom)
	{
		return true;
	}
	return bRet;

}

CSprite *CSprite::Clone()
{
	CSprite *pSprite = new CSprite(pRenderTarget,pTexture,vPos,width,height,tx,ty,vHotSpot,wsName);
	return pSprite;
}
void CSprite::SetEnabled(bool bVal)
{
	this->bEnabled = bVal;

}
bool CSprite::IsEnabled()
{
	return bEnabled;
}


